### ISSUE TITLE
Is it possible to update a segment's labelText property programmatically?

### ISSUE THREAD

**Issue by zachsa** [bbc/peaks.js#247]
I'm using React, and I initiate a segment with `labelText` as a reference to state within a component:

```
labelText: this.state.text
```

But updating the component state doesn't trigger the label of the segment to update. And if i retrieve the segment and update the labelText field then the changed label text doesn't reflect on the waveform.

Is it possible to force the labelText to redraw without re-rendering the waveform?

**Comment by zachsa** [comment]
I would be interested in working on this. Please let me know what I should look into and how to proceed. Thanks!

**Comment by chrisn** [comment]
Hi Zach,

Thank you! Here are a few notes, I hope these are helpful...

Segments are represented by [`Segment`](https://github.com/bbc/peaks.js/blob/master/src/main/markers/segment.js) objects, and these are managed by a [WaveformSegments](https://github.com/bbc/peaks.js/blob/master/src/main/markers/waveform.segments.js) object, which provides the `peaks.segments` API methods.

We want to be able to modify any of the segment properties and for those changes to be reflected in the UI.

The first step would be to modify `Segment` to store a reference to the `WaveformSegments` object, so we can notify it when things change:

```javascript
function Segment(parent, id, startTime, endTime, labelText, color, editable) {
  this._parent = parent; // leading underscore for private variables
  // etc
}
```

Then we can add methods to allow value changes. Rather than adding `getStartTime()` and `setStartTime()` methods we could use JavaScript [getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) and [setter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) syntax. This would also allow us to prevent changes to the `id` field, which must not be changed after the `Segment` is created.

The `WaveformSegments` object emits events to cause the UI to update. The UI is managed by the [`SegmentsLayer`](https://github.com/bbc/peaks.js/blob/master/src/main/views/segments-layer.js) object, which listens for those events.

We currently have `segments.add`, `segments.remove`, and `segments.remove_all` events. I suggest adding a `segments.update` event for segment value changes. The event handlers are set up in [`SegmentsLayer._registerEventHandlers()`](https://github.com/bbc/peaks.js/blob/master/src/main/views/segments-layer.js#L48).

The UI objects come from [Konva.js](https://konvajs.github.io/). Each segment is represented in the UI by a [`Konva.Group`](https://konvajs.github.io/api/Konva.Group.html). You can get the `Konva.Group` for a given `Segment` using this code within `SegmentsLayer`:

```javascript
var segmentGroup = this._segmentGroups[segment.id];
```

The `segmentGroup` has properties: `waveformShape`, `label`, `inMarker`, and `outMarker`.

The `inMarker` and `outMarker` come from [`createSegmentMarker`](https://github.com/bbc/peaks.js/blob/master/src/main/waveform.mixins.js#L35).

The idea of putting `createSegmentMarker` into the Peaks options was to allow user customisation of the UI. This part of the design could use some more thought, I don't think what we have now is ideal, and isn't documented. Actually, the whole UI appearance of points and segments could be improved, but that's a separate topic.

The `label` comes from [`createSegmentLabel`](https://github.com/bbc/peaks.js/blob/master/src/main/waveform.mixins.js#L139), again to allow user customisation. The label is a [`Konva.Text`](https://konvajs.github.io/api/Konva.Text.html) object, so changing the label text should be as simple as:

```javascript
segmentGroup.label.setText(newLabelText);
```

If you want to work on some of this, please create a branch rather than working directly on `master`.

**Comment by zachsa** [comment]
Hi @chrisn - thanks for the information. I'm going to see if I'm able to make a pull request on this.

**Comment by zachsa** [comment]
@chrisn

WRT to your suggestion of a `segments.update()` function; I initially took this to imply a new function on the `WaveformSegments` prototype with a signature something along the lines of the following:

```
WaveformSegments.prototype.update = function(id, {startTime, endTime, labelText, color, editable}) {
  var segment = this.getSegment(id)
  // Update whatever segment properties are set
  this._peaks.emit('segments.updated', segment)
  return segment
}
```

But then I realized that adding `this._parent = parent` to the `Segment` constructor I'm then able to emit peaks events from `segment` instances, in which case instead of having `WaveformSegments.prototype.update` I can envision instead the update function on the `Segment` constructor prototype:

```
Segment.prototype.update({startTime, endTime, labelText, color, editable}) {

  // Update segment properties
  if (startTime !== undefined) this.startTime = startTime;
  if (endTime !== undefined) this.endTime = endTime;
  if (labelText !== undefined) this.labelText = labelText;
  if (color !== undefined) this.color = color;
  if (editable !== undefined) this.editable = editable;

  // Trigger UI updates
  this._parent._peaks.emit('segments.updated', segment);

  // Return updated segment
  return this;
}
```

But I'm not sure why you mention getters/setters specifically?

> Rather than adding getStartTime() and setStartTime() methods we could use JavaScript getter and setter syntax. This would also allow us to prevent changes to the id field, which must not be changed after the Segment is created.

With the `Segment` constructor defined like this (specifically how startTime/endTIme is currently defined on `Segment` instances):

```
function Segment(...) {
  this.startTime = startTime;
  this.endTIme  = endTime;
  ...
}
```

I'm not sure how to define getters/setters in such a way that makes sense with properties set on the `this` object from within the constructor function. Would you mind giving me an example of what you mean?

I also don't think it would be possible to update a segment's `id` from within the `Segment.prototype.update` function I've mentioned above - so I may also be missing something with regards to this.

Sorry if I'm way off in my understanding of the code base! If so, please let me know - I'd really like to make this pull request as it will be my first contribution to an open source project.

**Comment by chrisn** [comment]
Thanks for looking into this, I really appreciate it!

In my previous comment I wasn't suggesting a `segments.update` function per se. My suggestion was for an event that would be emitted whenever a segment attribute changes (which is `segments.updated` in your example).

About the getters and setters, the idea was to enable us to write code like this:
```javascript
segment.startTime = 15.0; // emits segments.update event
```
in preference to this:
```javascript
segment.setStartTime(15.0);
```
There's more info on the approach [here](https://stackoverflow.com/questions/812961). The implementation would be something like this:
```javascript
Object.defineProperty(Segment.prototype, 'startTime', {
  get: function() {
    return this._startTime;
  },

  set: function(startTime) {
    this._startTime = startTime;
    this._parent._peaks.emit('segments.update', this);
  }
});
```
We could prevent changes to `id` by simply not providing a setter for it.

But, the `update` method you're suggesting is interesting. Which of these do you prefer?

**Comment by zachsa** [comment]
Ah i see!

I wasn't familiar enough with property descriptors to fully understand what you were suggesting. I like this approach and will do it that way.

**Comment by zachsa** [comment]
Hi @chrisn.

I have a working example of updating segment labels automatically. It's achieved by:

1. Passing `new Segment(...)` a reference to the current `WaveformSegments` context (instance). So, in the `waveform.segments.js`, in the `WaveformSegments.prototype._createSegment` function:
```
var segment = new Segment(
  this,
  Utils.isNullOrUndefined(options.id) ? this._getNextSegmentId() : options.id,
  ...
)
```

2. Adjusting the Segment constructor to set private properties when called (in `segments.js`):
```
function Segment(...) {
  Object.defineProperties(this, {
    _parent: { value: parent, writable: false },
    _id: { value: id, writable: false },
    _startTime: { value: startTime, writable: true },
    ...
  });
}
```
And then specifying a public API to access these private properties via a `getter/setter`- based API. The setters trigger the `segments.update` events
```
Object.defineProperties(Segment.prototype, {
  parent: { get: function() { return this._parent; } },
  id: { get: function() { return this._id; } },
  startTime: {
    get: function() { return this._startTime; },
    set: function(val) {
      this._startTime = val;
      this._parent._peaks.emit('segments.update', this);
    }
  },
  ...
}
```

3. Added an event handler that updates the segments group in `SegmentsLayer.prototype._registerEventHandlers` that updates the text:
```
this._peaks.on('segments.update', function(segment) {
  var segmentGroup = self._segmentGroups[segment.id];
  segmentGroup.label.setText(segment.labelText);
  ... I know that startTime and endTime are already updatable. Should any other properties be updated here?
});
```

As a means of testing, I adjusted `index.html`:

1. `renderSegments` function outputs a text `<input oninput="segmentUpdated(...)/>` for the label within the `<td/>` element
2. I added a `window.segmentUpdated` function to handle these events and use the new `Segment` API

Do you have any feedback on the above? In terms of a pull request, what would be ideal in terms of extending the test suite? (I haven't looked at this at all).

====================

On a separate note, I opened https://github.com/bbc/peaks.js/issues/250, because I couldn't select segments.

I see that a bunch of event handlers are setup in `SegmentsLayer.prototype._createSegmentGroup()`. It takes a few lines of code to emit events when a segment is interacted with. i.e.: 

```
segmentGroup.waveformShape.on('click',  function(event) {
  ...
  self._peaks.emit('segments.clicked', event.target.segment) // I see that the clicked segment is identifyable
}
```
And emitting similar events in the `mouseenter` and `mouseleave` event handlers in this same function.

This, in combination with being able to update segment colors allows for 'selecting' segments

**Comment by chrisn** [comment]
Thanks @zachsa! I think I'd prefer to have separate pull requests for the segment selection change.

I would only use `Object.defineProperties` for the getters and setters on the `Segment` attributes, and not for the private properties, I don't think the extra code is really needed. That is, I'd stick with:

```javascript
function Segment(id, startTime, endTime, labelText, color, editable) {
  this._id = id;
  // etc
}
```

We'll also need to think about validation. It shouldn't be possible to set a negative start or end time, or put the start time after the end time. We currently have this in [`WaveformSegments._createSegment`](https://github.com/bbc/peaks.js/blob/master/src/main/markers/waveform.segments.js#L121).

The rest of the changes you describe sound good to me.

For testing, I would add test cases [here](https://github.com/bbc/peaks.js/blob/master/test/unit/segment-spec.js). For each attribute, I would write tests for valid and invalid changes. Valid changes should result in a `segments.update` event, whereas invalid changes should throw an error, not emit an event, and leave the segment unmodified.

Doing this tests that you can modify the attributes, but does not test that the display is updated. We don't have too many tests for that right now. Those we do have work by spying on calls to Konva.js, which don't actually check that the rendering is correct, just that *something* got redrawn - see [here](https://github.com/bbc/peaks.js/blob/master/test/unit/waveform-segments-spec.js) for some examples.

I'd like to improve this aspect of the testing, so any ideas are welcome!

Feel free to open a pull request and I'll offer review feedback as we go.

**Comment by zachsa** [comment]
Sorry - I have been a little occupied. But I'm back! Will open a pull request shortly

**Comment by zachsa** [comment]
@chrisn - The commit is to a fork of the repo on my master branch. Would you mind having a look? Let me know what the best way to create a pull request is - i.e. would you like me to check out a branch and then do a pull request from there?

Something in this code (`views/segments-layer.js`):

```js
this._peaks.on('segments.update', function(segment) {
  var segmentGroup = self._segmentGroups[segment.id];

  segmentGroup.label.setText(segment.labelText);
});
```
when I try to set the label text of a segment that isn't 'seen' (i.e. if the segment is out of the window range, and I try to set it, then this error occurs):
```
segments-layer.js:19 Uncaught TypeError: Cannot read property 'label' of undefined
    at Peaks.<anonymous> (segments-layer.js:19)
    at Peaks.EventEmitter.emit (eventemitter2.js:384)
    at Segment.set (segment.js:67)
    at window.segmentUpdated ((index):186)
    at HTMLInputElement.oninput ((index):1)
```

**Comment by zachsa** [comment]
@chrisn - The issue that I mentioned is fixed. Next up - some basic tests!

**Comment by zachsa** [comment]
I've actually broken some tests - will fix before I comment again

**Comment by chrisn** [comment]
Many thanks! I'd prefer to receive a pull request on a feature branch, rather than master, just to avoid the risk of the master branches diverging.

**Comment by zachsa** [comment]
> Thanks @zachsa! I think I'd prefer to have separate pull requests for the segment selection change.
> 
> I would only use `Object.defineProperties` for the getters and setters on the `Segment` attributes, and not for the private properties, I don't think the extra code is really needed. That is, I'd stick with:
> 
> ```js
> function Segment(id, startTime, endTime, labelText, color, editable) {
>   this._id = id;
>   // etc
> }
> ```
> 
> We'll also need to think about validation. It shouldn't be possible to set a negative start or end time, or put the start time after the end time. We currently have this in [`WaveformSegments._createSegment`](https://github.com/bbc/peaks.js/blob/master/src/main/markers/waveform.segments.js#L121).
> 
> The rest of the changes you describe sound good to me.
> 
> For testing, I would add test cases [here](https://github.com/bbc/peaks.js/blob/master/test/unit/segment-spec.js). For each attribute, I would write tests for valid and invalid changes. Valid changes should result in a `segments.update` event, whereas invalid changes should throw an error, not emit an event, and leave the segment unmodified.
> 
> Doing this tests that you can modify the attributes, but does not test that the display is updated. We don't have too many tests for that right now. Those we do have work by spying on calls to Konva.js, which don't actually check that the rendering is correct, just that _something_ got redrawn - see [here](https://github.com/bbc/peaks.js/blob/master/test/unit/waveform-segments-spec.js) for some examples.
> 
> I'd like to improve this aspect of the testing, so any ideas are welcome!
> 
> Feel free to open a pull request and I'll offer review feedback as we go.

Question (if you don't mind me asking). What's the benefit of leaving the internal properties exposed? i.e., looking at the test suite, I see that looking at the keys in segment objects is quite messy vs if the 'internal' properties where defined via property descriptors they wouldn't be visible at all.

**Comment by chrisn** [comment]
I think I'd need to see an example from the test suite to comment.

**Comment by zachsa** [comment]
Apologies - I only half explained the context.

Looking at the `api-segments-spec.js` tests, I found that [THESE](https://github.com/bbc/peaks.js/blob/master/test/unit/api-segments-spec.js#L164-L165) tests were failing. i.e.
```
...
expect(p.segments.getSegments()[0]).to.include.keys('startTime', 'endTime');
...
```

Previously `startTime` was defined when the `Segment` constructor was called: `this.startTime = startTime` ([HERE](https://github.com/bbc/peaks.js/blob/master/src/main/markers/segment.js#L29))

But I'm changing that to `this._startTime = startTime`. And then I extend the `Segment` prototype object to include `statTime` etc keys.

I assume this is why the `expect` is failing - because the `startTime` key is no longer a segment-instance's 'own' property.

Now... when I logged the segment to look at the keys, it was a quite messy since all the 'private', or at least keys demarcated as private via an underscore, are not actually private but still visible. It's possible to define private, non-enumerable keys that could clean up logs when logging objects. This is what I mentioned above: 

```
function Segment(...) {
  Object.defineProperties(this, {
    _parent: { value: parent, writable: false },
    _id: { value: id, writable: false },
    _startTime: { value: startTime, writable: true },
    ...
  });
}
```

So I was wondering what/if there are any cons are to this approach that I don't know of? (I agree that it makes for less readable code, and therefore is better to avoid in general).

**Comment by zachsa** [comment]
The pull request for this issue is here: https://github.com/bbc/peaks.js/pull/260

**Comment by chrisn** [comment]
That test is simply checking that `segments.add` actually adds the segments you give it. So instead of checking for the keys, I'd rewrite the test as:

```javascript
expect(p.segments.getSegments()[0].startTime).to.equal(0);
expect(p.segments.getSegments()[0].endTime).to.equal(10);
expect(p.segments.getSegments()[1].startTime).to.equal(5);
expect(p.segments.getSegments()[1].endTime).to.equal(10);
```
This way, the test is verifying the public interface rather than implementation details.

**Comment by chrisn** [comment]
Resolved via #260.
